pkg/scale/decode.go

```diff
@@ -0,0 +1,539 @@
+package scale
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/big"
+	"reflect"
+)
+
+func Unmarshal(data []byte, dst interface{}) (err error) {
+	dstv := reflect.ValueOf(dst)
+	if dstv.Kind() != reflect.Ptr || dstv.IsNil() {
+		err = fmt.Errorf("unsupported dst: %T, must be a pointer to a destination", dst)
+		return
+	}
+
+	buf := &bytes.Buffer{}
+	ds := decodeState{}
+	_, err = buf.Write(data)
+	if err != nil {
+		return
+	}
+	ds.Buffer = *buf
+	err = ds.unmarshal(dstv.Elem())
+	if err != nil {
+		return
+	}
+	return
+}
+
+type decodeState struct {
+	bytes.Buffer
+}
+
+func (ds *decodeState) unmarshal(dstv reflect.Value) (err error) {
+	in := dstv.Interface()
+	switch in.(type) {
+	case *big.Int:
+		err = ds.decodeBigInt(dstv)
+	case *Uint128:
+		err = ds.decodeUint128(dstv)
+	case int, uint:
+		err = ds.decodeUint(dstv)
+	case int8, uint8, int16, uint16, int32, uint32, int64, uint64:
+		err = ds.decodeFixedWidthInt(dstv)
+	case []byte:
+		err = ds.decodeBytes(dstv)
+	case string:
+		err = ds.decodeBytes(dstv)
+	case bool:
+		err = ds.decodeBool(dstv)
+	default:
+		switch reflect.TypeOf(in).Kind() {
+		case reflect.Ptr:
+			err = ds.decodePointer(dstv)
+		case reflect.Struct:
+			t := reflect.TypeOf(in)
+			// check if this is a convertible to Result, if so encode using decodeResult
+			switch t.ConvertibleTo(reflect.TypeOf(Result{})) {
+			case true:
+				err = ds.decodeResult(dstv)
+			case false:
+				err = ds.decodeStruct(dstv)
+			}
+		case reflect.Array:
+			err = ds.decodeArray(dstv)
+		case reflect.Slice:
+			t := reflect.TypeOf(in)
+			// check if this is a convertible to VaryingDataType, if so encode using encodeVaryingDataType
+			switch t.ConvertibleTo(reflect.TypeOf(VaryingDataType{})) {
+			case true:
+				err = ds.decodeVaryingDataType(dstv)
+			case false:
+				err = ds.decodeSlice(dstv)
+			}
+		default:
+			_, ok := in.(VaryingDataTypeValue)
+			switch ok {
+			case true:
+				var temp reflect.Value
+				t := reflect.TypeOf(in)
+				switch t.Kind() {
+				// TODO: support more primitive types.  Do we need to support arrays and slices as well?
```

Comment:
is this for values within a `VaryingDataType`? if so, yes, I think the current usages for variadic types are are `common.Hash` or `uint64` type in the `lib/common/variadic` package

Reply:
I'll add those types.